gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+@VisibleForTesting
+class GoogleCloudStorageClientReadChannel implements SeekableByteChannel {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private StorageResourceId resourceId;
+  private BlobId blobId;
+  private GoogleCloudStorageReadOptions readOptions;
+  private Storage storage;
+  // The size of this object generation, in bytes.
+  private final long objectSize;
+  @VisibleForTesting boolean randomAccess;
+  // True if this channel is open, false otherwise.
+  private boolean channelIsOpen = true;
+
+  // Current position in this channel, it could be different from contentChannelPosition if
+  // position(long) method calls were made without calls to read(ByteBuffer) method.
+  private long currentPosition = 0;
+  private ReadableByteChannel contentReadChannel = null;
+  // This is the actual current position in `contentChannel` from where read can happen.
+  // This remains unchanged of position(long) method call.
+  private long contentChannelPosition = -1;
+  private long contentChannelEnd = -1;
+  private int footerSize;
+
+  // Used as scratch space when reading bytes just to discard them when trying to perform small
+  // in-place seeks.
+  private byte[] skipBuffer = null;
+  // Size of buffer to allocate for skipping bytes in-place when performing in-place seeks.
+  static final int SKIP_BUFFER_SIZE = 8192;
+
+  // Prefetched footer content.
+  private byte[] footerContent;
+
+  public GoogleCloudStorageClientReadChannel(
+      Storage gcs,
+      GoogleCloudStorageItemInfo itemInfo,
+      @Nonnull GoogleCloudStorageReadOptions readOptions)
+      throws IOException {
+    validate(itemInfo);
+    this.storage = gcs;
+    this.resourceId =
+        new StorageResourceId(
+            itemInfo.getBucketName(), itemInfo.getObjectName(), itemInfo.getContentGeneration());
+    this.blobId =
+        BlobId.of(
+            resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+    this.readOptions = readOptions;
+    this.objectSize = itemInfo.getSize();
+    this.randomAccess = readOptions.getFadvise().equals(Fadvise.RANDOM) ? true : false;
+    this.footerSize = (int) readOptions.getMinRangeRequestSize();
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    throwIfNotOpen();
+
+    // Don't try to read if the buffer has no space.
+    if (dst.remaining() == 0) {
+      return 0;
+    }
+
+    logger.atFiner().log(
+        "Reading %d bytes at %d position from '%s'", dst.remaining(), currentPosition, resourceId);
+
+    if (currentPosition == objectSize) {
+      return -1;
+    }
+
+    int totalBytesRead = 0;
+    while (dst.hasRemaining()) {
+      // if in Footer and not readRandom
```

Comment:
Is this comment accurate?

Reply:
No. it was for reference and got missed in cleanup. Will remove.